File: base/src/main/java/org/arend/proof/ProofSearchQuery.java
package org.arend.proof;

import org.arend.ext.util.Pair;
import org.arend.util.Range;

import java.util.*;
import java.util.function.Function;
import java.util.regex.MatchResult;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ProofSearchQuery {
    public final List<ProofSearchJointPattern> parameters;
    public final ProofSearchJointPattern codomain;

    public ProofSearchQuery(List<ProofSearchJointPattern> parameters, ProofSearchJointPattern codomain) {
        this.parameters = parameters;
        this.codomain = codomain;
    }

    public static class ProofSearchJointPattern {
        public final List<PatternTree> patterns;

        public ProofSearchJointPattern(List<PatternTree> patterns) {
            this.patterns = patterns;
        }

        public List<String> getAllIdentifiers() {
            List<String> result = new ArrayList<>();
            for (PatternTree pattern : patterns) {
                result.addAll(pattern.getAllIdentifiers());
            }
            return result;
        }

        @Override
        public String toString() {
            List<String> result = new ArrayList<>();
            for (PatternTree pattern : patterns) {
                result.add(pattern.toString());
            }
            return "<" + String.join("  \\and  ", result) + ">";
        }
    }

    public static class Token {
        public final String repr;
        public final Range<Integer> range;

        public Token(String repr, Range<Integer> range) {
            this.repr = repr;
            this.range = range;
        }
    }

    public interface ParsingResult<T> {
        class OK<T> implements ParsingResult<T> {
            public final T value;
            public OK(T value) { this.value = value; }
        }
        class Error<T> implements ParsingResult<T> {
            public final String message;
            public final Range<Integer> range;
            public Error(String message, Range<Integer> range) { this.message = message; this.range = range; }
            @SuppressWarnings("unchecked")
            public <R> Error<R> cast() { return (Error<R>) this; }
        }

        default <U> ParsingResult<U> bind(Function<T, ParsingResult<U>> f) {
            if (this instanceof OK<T> ok) {
                return f.apply(ok.value);
            } else {
                return ((Error<T>) this).cast();
            }
        }

        default <U> ParsingResult<U> map(Function<T, U> f) {
            return bind(value -> new OK<>(f.apply(value)));
        }
    }

    public static ParsingResult<ProofSearchQuery> fromString(String pattern) {
        Pattern patternString = Pattern.compile("[^(){}\\s]+|\\(|\\)|\\{|}");
        Matcher matcher = patternString.matcher(pattern);
        List<MatchResult> matches = new ArrayList<>();
        while (matcher.find()) {
            matches.add(matcher.toMatchResult());
        }
        List<Token> tokens = matches.stream().map(match -> new Token(match.group(), new Range<>(match.start(), match.end()))).toList();
        return parseTokens(tokens);
    }

    private static ParsingResult<ProofSearchQuery> parseTokens(List<Token> tokens) {
        ParsingResult<int[]> braceMatchingResult = computeBraceMatching(tokens);
        if (braceMatchingResult instanceof ParsingResult.Error) return ((ParsingResult.Error<int[]>) braceMatchingResult).cast();
        int[] braceMatcher = ((ParsingResult.OK<int[]>) braceMatchingResult).value;

        List<Pair<Integer, Integer>> disjunctionBoundaries = splitTokens(tokens, "->");

        List<ParsingResult<ProofSearchJointPattern>> result = new ArrayList<>();
        for (Pair<Integer, Integer> disjunctionBoundary : disjunctionBoundaries) {
            Integer start = disjunctionBoundary.proj1;
            Integer end = disjunctionBoundary.proj2;
            List<Token> subTokens = tokens.subList(start, end);
            List<Pair<Integer, Integer>> disjunctionSubTokens = splitTokens(subTokens, "\\and");
            result.add(swap(disjunctionSubTokens.stream().map(pair ->
                    doParseTokens(pair.proj1 + start, pair.proj2 + start, tokens, braceMatcher)
            ).toList()).map(ProofSearchJointPattern::new));
        }
        ProofSearchQuery.ParsingResult<List<ProofSearchQuery.ProofSearchJointPattern>> rawPatterns = swap(result);
        return rawPatterns.map(proofSearchJointPatterns ->
                new ProofSearchQuery(proofSearchJointPatterns.subList(0, proofSearchJointPatterns.size() - 1), proofSearchJointPatterns.getLast()));
    }

    private static List<Pair<Integer, Integer>> splitTokens(List<Token> tokens, String pattern) {
        List<Pair<Integer, Integer>> newList = new ArrayList<>();
        int firstIndex = 0;
        for (int idx = 0; idx < tokens.size(); idx++) {
            if (tokens.get(idx).repr.equals(pattern)) {
                newList.add(new Pair<>(firstIndex, idx));
                firstIndex = idx + 1;
            }
        }
        newList.add(new Pair<>(firstIndex, tokens.size()));
        return newList;
    }

    public static <T> ParsingResult<List<T>> swap(List<ParsingResult<T>> list) {
        ParsingResult<List<T>> result = new ParsingResult.OK<>(new ArrayList<>());
        for (ParsingResult<T> t : list) {
            result = result.bind(collectedList ->
                    t.map(e -> {
                        collectedList.add(e);
                        return collectedList;
                    })
            );
        }
        return result;
    }

    static ParsingResult<PatternTree> doParseTokens(int position, int limit, List<Token> tokens, int[] braceMatcher) {
        if (position == limit) {
            Range range;
            if (tokens.isEmpty()) {
                range = new Range<>(0, 0);
            } else if (limit == tokens.size()) {
                range = tokens.get(position - 1).range;
            } else {
                range = tokens.get(position).range;
            }
            return new ParsingResult.Error<>("Unexpected end of input", range);
        }

        List<Pair<PatternTree, PatternTree.Implicitness>> nodes = new ArrayList<>();
        int currentPosition = position;

        while (currentPosition != limit) {
            String tokenRepr = tokens.get(currentPosition).repr;

            switch (tokenRepr) {
                case "_":
                    nodes.add(new Pair<>(PatternTree.Wildcard.INSTANCE, PatternTree.Implicitness.EXPLICIT));
                    break;

                case "(": {
                    int closingBrace = braceMatcher[currentPosition];
                    ParsingResult<PatternTree> result = doParseTokens(currentPosition + 1, closingBrace, tokens, braceMatcher);

                    if (result instanceof ParsingResult.OK) {
                        PatternTree value = ((ParsingResult.OK<PatternTree>) result).value;
                        nodes.add(new Pair<>(value, PatternTree.Implicitness.EXPLICIT));
                        currentPosition = closingBrace;
                    } else {
                        return result;
                    }
                    break;
                }

                case ")":
                    return new ParsingResult.Error<>("Unexpected ')'", tokens.get(currentPosition).range);

                case "{": {
                    int closingBrace = braceMatcher[currentPosition];
                    ParsingResult<PatternTree> result = doParseTokens(currentPosition + 1, closingBrace, tokens, braceMatcher);

                    if (result instanceof ParsingResult.OK) {
                        PatternTree value = ((ParsingResult.OK<PatternTree>) result).value;
                        nodes.add(new Pair<>(value, PatternTree.Implicitness.IMPLICIT));
                        currentPosition = closingBrace;
                    } else {
                        return result;
                    }
                    break;
                }

                case "}":
                    return new ParsingResult.Error<>("Unexpected '}'", tokens.get(currentPosition).range);

                default:
                    nodes.add(new Pair<>(
                            new PatternTree.LeafNode(Arrays.asList(tokenRepr.split("\\."))),
                            PatternTree.Implicitness.EXPLICIT
                    ));
                    break;
            }
            currentPosition += 1;
        }

        if (nodes.size() == 1 && nodes.getFirst().proj2 != PatternTree.Implicitness.IMPLICIT) {
            return new ParsingResult.OK<>(nodes.getFirst().proj1);
        } else {
            return new ParsingResult.OK<>(new PatternTree.BranchingNode(nodes));
        }
    }

    private static ParsingResult<int[]> computeBraceMatching(List<Token> tokens) {
        int[] array = new int[tokens.size()];
        List<Pair<Character, Integer>> stack = new ArrayList<>();

        for (int idx = 0; idx < tokens.size(); idx++) {
            String repr = tokens.get(idx).repr;
            switch (repr) {
                case "->":
                    if (!stack.isEmpty()) {
                        return new ParsingResult.Error<>("'->' is allowed only on top level", tokens.get(idx).range);
                    }
                    break;
                case "\\and":
                    if (stack.size() > 1) {
                        return new ParsingResult.Error<>("'\\and' is allowed only in clauses", tokens.get(idx).range);
                    }
                    break;
                case "(":
                    stack.add(new Pair<>(')', idx));
                    break;
                case "{":
                    stack.add(new Pair<>('}', idx));
                    break;
                case ")": {
                    if (stack.isEmpty()) {
                        return new ParsingResult.Error<>("Unexpected ')'", tokens.get(idx).range);
                    }

                    Pair<Character, Integer> top = stack.removeLast();
                    char topBrace = top.proj1;
                    int topIndex = top.proj2;

                    if (topBrace != ')') {
                        return new ParsingResult.Error<>("Unexpected ')'", tokens.get(idx).range);
                    }
                    array[topIndex] = idx;
                    break;
                }

                case "}": {
                    if (stack.isEmpty()) {
                        return new ParsingResult.Error<>("Unexpected '}'", tokens.get(idx).range);
                    }

                    Pair<Character, Integer> top = stack.removeLast();
                    char topBrace = top.proj1;
                    int topIndex = top.proj2;

                    if (topBrace != '}') {
                        return new ParsingResult.Error<>("Unexpected '}'", tokens.get(idx).range);
                    }

                    array[topIndex] = idx;
                    break;
                }
            }
        }

        if (!stack.isEmpty()) {
            Pair<Character, Integer> first = stack.getFirst();
            return new ParsingResult.Error<>("Could not find a matching brace", tokens.get(first.proj2).range);
        } else {
            return new ParsingResult.OK<>(array);
        }
    }

    public List<String> getAllIdentifiers() {
        List<String> result = new ArrayList<>();
        for (ProofSearchJointPattern p : parameters) result.addAll(p.getAllIdentifiers());
        result.addAll(codomain.getAllIdentifiers());
        return result;
    }

    @Override
    public String toString() {
        return String.join(" -> ", parameters.stream().map(ProofSearchJointPattern::toString).toList()) +
                (parameters.isEmpty() ? "" : " --> ") + codomain.toString();
    }
}

File: intellij/src/main/kotlin/org/arend/search/proof/ProofSearchQuery.kt
package org.arend.search.proof

import com.intellij.util.SmartList
import java.util.regex.Pattern

data class ProofSearchQuery(val parameters: List<ProofSearchJointPattern>, val codomain: ProofSearchJointPattern) {
    companion object {
        fun fromString(pattern: String) : ParsingResult<ProofSearchQuery> = patternToQuery(pattern)
    }
    fun getAllIdentifiers() = (parameters + codomain).flatMap(ProofSearchJointPattern::getAllIdentifiers)

    override fun toString(): String {
        return parameters.joinToString(" -> ") { it.toString() } + (if (parameters.isEmpty()) "" else " --> ") + codomain.toString()
    }
}

@JvmInline
value class ProofSearchJointPattern(val patterns: List<PatternTree>) {
    fun getAllIdentifiers() = patterns.flatMap(PatternTree::getAllIdentifiers)

    override fun toString(): String {
        return patterns.joinToString("  \\and  ", "<", ">") { it.toString() }
    }
}

sealed interface PatternTree {
    enum class Implicitness {
        IMPLICIT, EXPLICIT;

        fun compactLBrace(): String = when (this) {
            IMPLICIT -> "{"
            EXPLICIT -> ""
        }

        fun compactRBrace(): String = when (this) {
            IMPLICIT -> "}"
            EXPLICIT -> ""
        }

        fun toBoolean(): Boolean = this == EXPLICIT
    }

    @JvmInline
    value class BranchingNode(val subNodes: List<Pair<PatternTree, Implicitness>>) : PatternTree {
        override fun getAllIdentifiers(): List<String> = subNodes.flatMap { it.first.getAllIdentifiers() }

        override fun toString(): String =
            subNodes.joinToString(" ", "[", "]", transform = { "${it.second.compactLBrace()}${it.first}${it.second.compactRBrace()}" })
    }

    @JvmInline
    value class LeafNode(val referenceName: List<String>) : PatternTree {
        override fun getAllIdentifiers(): List<String> = listOf(referenceName.last())

        override fun toString(): String = referenceName.joinToString(".")
    }

    object Wildcard : PatternTree {
        override fun getAllIdentifiers(): List<String> = emptyList()

        override fun toString(): String = "_"
    }

    fun getAllIdentifiers() : List<String>
}

private typealias Token = String

private fun patternToQuery(pattern: String): ParsingResult<ProofSearchQuery> {
    val tokens: List<Token> =
        Pattern.compile("""[^(){}\p{Space}]+|\(|\)|\{|}""").toRegex().findAll(pattern).map { it.value }.toList()
    return parseTokens(tokens)
}

sealed interface ParsingResult<out T> {

    data class OK<T>(val value: T) : ParsingResult<T>

    data class Error<T>(val message: String, val offset: Int) : ParsingResult<T> {
        @Suppress("UNCHECKED_CAST")
        fun <R> cast(): Error<R> = this as Error<R>
    }

    fun <U> bind(f: (T) -> ParsingResult<U>): ParsingResult<U> = when (this) {
        is OK -> f(value)
        is Error -> cast()
    }

    fun <U> map(f: (T) -> U): ParsingResult<U> = bind { OK(f(it)) }
}

private fun parseTokens(tokens: List<Token>): ParsingResult<ProofSearchQuery> {

    val braceMatchingResult = computeBraceMatching(tokens)
    if (braceMatchingResult is ParsingResult.Error) return braceMatchingResult.cast()
    val braceMatcher = (braceMatchingResult as ParsingResult.OK).value

    fun doParseTokens(position: Int, limit: Int): ParsingResult<PatternTree> {
        if (position == limit) return ParsingResult.Error("Unexpected end of input", position)
        val nodes: MutableList<Pair<PatternTree, PatternTree.Implicitness>> = mutableListOf()
        var currentPosition = position

        while (currentPosition != limit) {
            when (val token = tokens[currentPosition]) {
                "_" -> nodes.add(PatternTree.Wildcard to PatternTree.Implicitness.EXPLICIT)
                "(" -> {
                    val closingBrace = braceMatcher[currentPosition]
                    when (val result = doParseTokens(currentPosition + 1, closingBrace)) {
                        is ParsingResult.OK -> {
                            nodes.add(result.value to PatternTree.Implicitness.EXPLICIT)
                            currentPosition = closingBrace
                        }
                        is ParsingResult.Error -> return result
                    }
                }
                ")" -> return ParsingResult.Error("Unexpected ')'", currentPosition)
                "{" -> {
                    val closingBrace = braceMatcher[currentPosition]
                    when (val result = doParseTokens(currentPosition + 1, closingBrace)) {
                        is ParsingResult.OK -> {
                            nodes.add(result.value to PatternTree.Implicitness.IMPLICIT)
                            currentPosition = closingBrace
                        }
                        is ParsingResult.Error -> return result
                    }
                }
                "}" -> return ParsingResult.Error("Unexpected '}'", currentPosition)
                else -> nodes.add(PatternTree.LeafNode(token.split(".")) to PatternTree.Implicitness.EXPLICIT)
            }
            currentPosition += 1
        }
        return if (nodes.size == 1 && nodes[0].second != PatternTree.Implicitness.IMPLICIT) {
            ParsingResult.OK(nodes[0].first)
        } else {
            ParsingResult.OK(PatternTree.BranchingNode(nodes))
        }
    }

    val disjunctionBoundaries = tokens.split("->")

    val rawPatterns = disjunctionBoundaries.map { (start, end) ->
        tokens
            .subList(start, end)
            .split("\\and")
            .map { doParseTokens(start + it.first, start + it.second) }
            .swap()
            .map(::ProofSearchJointPattern)
    }
        .swap()
    return rawPatterns.map {
        ProofSearchQuery(it.subList(0, it.lastIndex), it.last())
    }
}

private fun <T> List<T>.split(t: T): List<Pair<Int, Int>> {
    val newList = SmartList<Pair<Int, Int>>(SmartList())
    var firstIndex = 0
    for (idx in this.indices) {
        if (this[idx] == t) {
            newList.add(firstIndex to idx)
            firstIndex = idx + 1
        }
    }
    newList.add(firstIndex to this.size)
    return newList
}

private fun <T> List<ParsingResult<T>>.swap() : ParsingResult<List<T>> {
    return fold(ParsingResult.OK(SmartList<T>()) as ParsingResult<SmartList<T>>) { collector, t -> collector.bind { list -> t.map { e -> list.also { it.add(e) } } }}
}

private fun computeBraceMatching(tokens: List<Token>): ParsingResult<Array<Int>> {
    val array = Array(tokens.size) { 0 }
    val stack = mutableListOf<Pair<Char, Int>>()
    for (idx in tokens.indices) {
        when (tokens[idx]) {
            "->" -> if (stack.isNotEmpty()) return ParsingResult.Error("'->' is allowed only on top level", idx)
            "\\and" -> if (stack.size > 1) return ParsingResult.Error("'\\and' is allowed only in clauses", idx)
            "(" -> stack.add(')' to idx)
            "{" -> stack.add('}' to idx)
            ")" -> {
                val (topBrace, topIndex) = stack.removeLastOrNull() ?: return ParsingResult.Error("Unexpected ')'", idx)
                if (topBrace != ')') return ParsingResult.Error("Unexpected ')'", idx)
                array[topIndex] = idx
            }
            "}" -> {
                val (topBrace, topIndex) = stack.removeLastOrNull() ?: return ParsingResult.Error("Unexpected '}'", idx)
                if (topBrace != '}') return ParsingResult.Error("Unexpected '}'", idx)
                array[topIndex] = idx
            }
        }
    }
    return if (stack.isNotEmpty()) {
        ParsingResult.Error("Could not find a matching brace", stack.first().second)
    } else {
        ParsingResult.OK(array)
    }
}